### PR TITLE
Fixed #4282 -- Made startapp/startproject management commands honor umask.

### PR DESCRIPTION
**Pull request by awtimmering**
Co-authored-by: c.schmitt@briefdomain.de

### CODE REVIEW

**Comment by awtimmering** [comment]
For reference:
[Python issue #38417 Add support for settting umask in subprocess children](https://bugs.python.org/issue38417)

File: tests/admin_scripts/tests.py [hunk 1]
```diff
@@ -118,17 +119,26 @@ def run_test(self, args, settings_file=None, apps=None):
         test_environ['PYTHONPATH'] = os.pathsep.join(python_path)
         test_environ['PYTHONWARNINGS'] = ''
 
+        additional_kwargs = dict()
+        if umask and sys.hexversion >= 0x3090000:
+            # umask setting in subprocess.run() only supported from 3+
```
Comment (kezabelle) [inline tests/admin_scripts/tests.py:124]: `3+` should be `3.9+` if I'm reading the `skipIf` and `hexversion` correctly?

Comment (awtimmering) [reply tests/admin_scripts/tests.py:124]: Oops - absolutely right, fixing. Thanks for catching.

File: tests/admin_scripts/tests.py [hunk 2]
```diff
@@ -118,17 +119,26 @@ def run_test(self, args, settings_file=None, apps=None):
         test_environ['PYTHONPATH'] = os.pathsep.join(python_path)
         test_environ['PYTHONWARNINGS'] = ''
 
+        additional_kwargs = dict()
+        if umask and sys.hexversion >= 0x3090000:
```
Comment (felixxm) [inline tests/admin_scripts/tests.py:123]: Use `PY39` instead:
```suggestion
        if umask and PY39:
```

File: tests/admin_scripts/tests.py [hunk 3]
```diff
@@ -118,17 +119,26 @@ def run_test(self, args, settings_file=None, apps=None):
         test_environ['PYTHONPATH'] = os.pathsep.join(python_path)
         test_environ['PYTHONWARNINGS'] = ''
 
+        additional_kwargs = dict()
```
Comment (felixxm) [inline tests/admin_scripts/tests.py:122]: I would rename it to `kwargs`
```suggestion
        kwargs = {}
```

File: tests/admin_scripts/tests.py [hunk 4]
```diff
@@ -2297,6 +2307,24 @@ def test_custom_project_template_exclude_directory(self):
         not_excluded = os.path.join(testproject_dir, project_name)
         self.assertIs(os.path.exists(not_excluded), True)
 
+    @unittest.skipIf(os.name == 'nt', "platform doesn't support umask")
+    @unittest.skipIf(sys.hexversion < 0x3090000, "umask test not supported prior to 3.9")
```
Comment (felixxm) [inline tests/admin_scripts/tests.py:2311]:
```suggestion
    @unittest.skipUnless(PY39, 'subprocess.run()'s umask was added in Python 3.9')
```

File: tests/admin_scripts/tests.py [hunk 5]
```diff
@@ -2297,6 +2307,24 @@ def test_custom_project_template_exclude_directory(self):
         not_excluded = os.path.join(testproject_dir, project_name)
         self.assertIs(os.path.exists(not_excluded), True)
 
+    @unittest.skipIf(os.name == 'nt', "platform doesn't support umask")
```
Comment (felixxm) [inline tests/admin_scripts/tests.py:2310]:
```suggestion
    @unittest.skipIf(os.platform == 'win32', 'Python on Windows doesn't support umask.')
```

File: tests/admin_scripts/tests.py [hunk 6]
```diff
@@ -118,17 +119,26 @@ def run_test(self, args, settings_file=None, apps=None):
         test_environ['PYTHONPATH'] = os.pathsep.join(python_path)
         test_environ['PYTHONWARNINGS'] = ''
 
+        additional_kwargs = dict()
+        if umask and sys.hexversion >= 0x3090000:
+            # umask setting in subprocess.run() only supported from 3.9+
+            current_umask = os.umask(0)
+            os.umask(current_umask)
+            applied_umask = umask or current_umask
+            additional_kwargs['umask'] = applied_umask
```
Comment (felixxm) [inline tests/admin_scripts/tests.py:128]: Do we need to check a `current_mask`? This code is reachable only when `umask` is set.
```suggestion
            kwargs['umask'] = umask
```

File: django/core/management/templates.py [hunk 7]
```diff
@@ -345,6 +346,16 @@ def is_url(self, template):
         scheme = template.split(':', 1)[0].lower()
         return scheme in self.url_schemes
 
+    def apply_umask(self, old_path, new_path):
+        """Apply umask if operating system supports it."""
+        if not os.name == 'nt':
```
Comment (felixxm) [inline django/core/management/templates.py:351]: Should we check `posix` instead?
```suggestion
        if os.name == 'posix':
```

**Review by felixxm — COMMENTED** [review]
@awtimmering Thanks for this patch :+1: I left initial comments.

File: tests/admin_scripts/tests.py (hunk 6 above)
Comment (awtimmering) [reply tests/admin_scripts/tests.py:128]: Good point - unneeded leftover from a previous iteration.

File: django/core/management/templates.py (hunk 7 above)
Comment (awtimmering) [reply django/core/management/templates.py:351]: Agree, `os.name == 'posix'` makes sense and happy to change; in practice there seems to be no precedence in the code base for that with most exceptions being based on `sys.platform == 'win32'`.

**Review by claudep — APPROVED** [review]
I'm happy with other reviewers
Lazy is my second name!

**Review by felixxm — COMMENTED** [review]
@awtimmering Thanks for updates :+1: 

I pushed small edits.

**Pull request merged**